ODIN_II/SRC/generate_blif_stats.cpp

```diff
@@ -0,0 +1,1128 @@
+		{
+			if (global_args.high_level_block.provenance() == argparse::Provenance::SPECIFIED)
+				num_names++;
+			else
+			{
+				/*  Use the name of the driver pin as the name of the driver as long as that name is set, and is not equal to the name of the output pin.
+				 *  Otherwise, use the name of the driver node.*/
+				char *driver = node->input_pins[0]->net->driver_pin->name;
+				char *output = node->name;
+				if (!driver || !strcmp(driver,output))
+					driver = node->input_pins[0]->net->driver_pin->node->name; // I still don't know what this is
+				/* Skip if the driver and output have the same name (i.e. the output of a flip-flop) */
+				if (strcmp(driver,output) != 0)
+					num_names++;
+			}
+		}
+	}
+	float avg_output_pins_per_node = (float)output_pin_total / (float)num_nodes; // POs don't have output pins so probably < 1, not very interesting number
+	fprintf(out, "==============COUNTERS=============\n");
+	fprintf(out, "num_nodes: %u\n", num_nodes); // verified correct
+	fprintf(out, "num_latches: %u\n", num_latches); // verified correct
+	fprintf(out, "num_total_input_pins: %u\n", num_total_input_pins); // a measure of fanin to nodes
+	fprintf(out, "num_nets: %u\n", num_nets);
+	fprintf(out, "num_names: %u\n", num_names); // verified correct
+	fprintf(out, "num_edges_fo: %u\n", num_edges_fo); // All fanout from all nets
+	//fprintf(out, "num_edges_fi: %u\n", num_edges_fi); // probably pointless? should always be == num_total_input_pins
+	fprintf(out, "output_pin_total: %u\n", output_pin_total); // pointless? If including output nodes the number can be < 1
+	fprintf(out, "num_subckts: %u\n", num_subckts);
+	fprintf(out, "num_muxes: %u\n", num_muxes);
+	//fprintf(out, "num_ff_nodes: %d\n", netlist->num_ff_nodes); // # latches: This value is already available and seems correct always but we play it safe and count the latches
+	fprintf(out, "num_top_output_nodes: %u\n", netlist->num_top_output_nodes);	
+	fprintf(out, "num_input_nodes: %u\n", num_input_nodes); // this tends to be off by one compared to struct member, represents true input nodes, not clocks
+	if(verbose_mode)
+	{
+		fprintf(out, "(W/ clock %d top inputs)\n", netlist->num_top_input_nodes); // includes clock, not in parser
+		fprintf(out, "num_deadend_inputs: %u\n", num_deadend_inputs); // number of inputs that go nowhere, dist values of 0
+		fprintf(out, "num_clock_nodes: %u\n", num_clock_nodes); // this doesn't include global sim base clock which is confusingly of type input
+		fprintf(out, "num_logic_nodes: %u\n", num_logic_nodes);
+		fprintf(out, "num_carry_func: %u\n", num_carry_func);
+		fprintf(out, "num_generic_nodes: %u\n", num_generic_nodes);
+		fprintf(out, "num_unconn: %u\n", num_unconn); // unreliable, how would we find an unconn node?
+		fprintf(out, "num_memories: %u\n", num_memories);
+		fprintf(out, "num_adders: %u\n", num_adders);
+		fprintf(out, "num_deadends: %u\n", num_deadends); // unclear if fully reliable
```

Comment:
I would make sure it is reliable and later down the road a sanitize phase should clean this up